### ISSUE TITLE
Static and critical pod is evicted

### ISSUE THREAD

**Issue by Ladicle** [kubernetes/kubernetes#73572]
**What happened**:

Eviction manager evicts the below static and critical pod when the node observed MemoryPressure.
The k8s server version is v1.13.2, and the pod priority is enabled.

```yaml
Name:               node-prober-worker-7c8f45f1-zczzz-caas-m22qk-test
Namespace:          kube-system
Priority:           2000001000
PriorityClassName:  system-node-critical
...
Annotations:        kubernetes.io/config.source: file
...
Status:             Failed
Reason:             Evicted
Message:            The node was low on resource: memory. Container bph was using 11896Ki, which exceeds its request of 0. 
...
```

As far as I can see, the eviction manager does not evict such pods.
https://github.com/kubernetes/kubernetes/blob/6cea9266dcfb89d1d278d102885c03b11dbe26fc/pkg/kubelet/eviction/eviction_manager.go#L547

**What you expected to happen**:

Static pods with `SystemCriticalPriority` are not deleted as well as those with a critical pod annotation.

**How to reproduce it (as minimally and precisely as possible)**:

1. Boot pod with system-node-critical priority using by kubelet `--pod-manifest-path` option
2. Apply memory load to the target node until eviction manager observes memory pressure.
3. The critical and static pod is evicted.

**Anything else we need to know?**:

When I set critical-pod annotation (and enables `ExperimentalCriticalPodAnnotation`) insted of pod priority, the eviction manager **DOES NOT** evict the static and critical pod.

```
Jan 31 05:23:29 worker-7c8f45f1-zczzz-test-m22qk-test kubelet-wrapper[10083]: I0131 05:23:29.942701   10083 eviction_manager.go:358] eviction manager: pods ranked for eviction: node-prober-worker-7c8f45f1-zczzz-test-m22qk-test_kube-system(70eede23a9faee60311f7ee4a0d6457e), stress1-67f94bb9cf-m4qkx_kube-system(480601e2-2518-11e9-9b99-fa163efcff57), stress1-67f94bb9cf-72qxp_kube-system(0b77dd77-2518-11e9-9b99-fa163efcff57), stress1-67f94bb9cf-qtnt4_kube-system(0b7cc794-2518-11e9-9b99-fa163efcff57), stress1-67f94bb9cf-6fg72_kube-system(0b7ae76f-2518-11e9-9b99-fa163efcff57), stress1-67f94bb9cf-f7r8j_kube-system(0b7cd495-2518-11e9-9b99-fa163efcff57), stress1-67f94bb9cf-t5n9g_kube-system(49738ffc-2518-11e9-9b99-fa163efcff57), pod-network-problem-detector-tw9f6_kube-system(f0ba6ab2-2517-11e9-9b99-fa163efcff57), node-exporter-z49dm_kube-system(f0ba7978-2517-11e9-9b99-fa163efcff57), container-log-forwarder-s6z7s_kube-system(f6b56829-2517-11e9-9b99-fa163efcff57)
Jan 31 05:23:29 worker-7c8f45f1-zczzz-test-m22qk-test kubelet-wrapper[10083]: E0131 05:23:29.942727   10083 eviction_manager.go:548] eviction manager: cannot evict a critical static pod node-prober-worker-7c8f45f1-zczzz-test-m22qk-test_kube-system(70eede23a9faee60311f7ee4a0d6457e)
```

**Environment**:
- Kubernetes version (use `kubectl version`): v1.13.2
- Cloud provider or hardware configuration: baremetal

**Comment by Ladicle** [comment]
Right. It seems the expected behavior and the actual behavior are different. `ExperimentalCriticalPodAnnotation` will remove next k8s v1.14 (it's written in the document), but I'd like to wait for it until the critical priority can prevent eviction.

**Comment by relyt0925** [comment]
I see this as well. 
```
Feb 14 16:24:26 stage-dal10-crb6bbd7d143e14babbb7fa61470648230-w4.cloud.ibm kubelet.service[10075]: I0214 16:24:26.383983   10075 helpers.go:895] eviction manager: eviction criteria not yet met for threshold(signal=nodefs.available, operator=LessThan, value={<nil> 0.1}, gracePeriod=10m0s), duration: 1m29.703137346s
Feb 14 16:24:26 stage-dal10-crb6bbd7d143e14babbb7fa61470648230-w4.cloud.ibm kubelet.service[10075]: I0214 16:24:26.328774   10075 eviction_manager.go:400] eviction manager: pods ibm-master-proxy-static-10.177.180.213_kube-system(80da011274014ad7857773c0958a6493) successfully cleaned up
Feb 14 16:24:08 stage-dal10-crb6bbd7d143e14babbb7fa61470648230-w4.cloud.ibm kubelet.service[10075]: I0214 16:24:08.328177   10075 eviction_manager.go:187] eviction manager: pods ibm-master-proxy-static-10.177.180.213_kube-system(80da011274014ad7857773c0958a6493) evicted, waiting for pod to be cleaned up
Feb 14 16:24:08 stage-dal10-crb6bbd7d143e14babbb7fa61470648230-w4.cloud.ibm kubelet.service[10075]: I0214 16:24:08.328140   10075 eviction_manager.go:563] eviction manager: pod ibm-master-proxy-static-10.177.180.213_kube-system(80da011274014ad7857773c0958a6493) is evicted successfully
Feb 14 16:24:07 stage-dal10-crb6bbd7d143e14babbb7fa61470648230-w4.cloud.ibm kubelet.service[10075]: I0214 16:24:07.908949   10075 eviction_manager.go:358] eviction manager: pods ranked foreviction: ibm-master-proxy-static-10.177.180.213_kube-system(80da011274014ad7857773c0958a6493), ssh-daemonset-kskpw_default(2e0c2a58-305e-11e9-8d45-e280b8281f73), stress-tester-688zc_default(6f64788c-306b-11e9-8d45-e280b8281f73), calico-node-5j42w_kube-system(d7d3c60b-3055-11e9-8d45-e280b8281f73), ibm-kube-fluentd-pkpkx_kube-system(d7cf8425-3055-11e9-8d45-e280b8281f73), ibm-keepalived-watcher-cfjvq_kube-system(d7d5d2c3-3055-11e9-8d45-e280b8281f73)
```

Start from the bottom and read up. It ranks ibm-master-proxy-static which is a static pod that has priority system-node-critical. It actually is always ranked first as well.

**Comment by relyt0925** [comment]
```
Feb 14 19:30:31 dev-dal10-cr41b86b475ced4687a7ef34b5dc0513e4-w1 kubelet.service[7498]: I0214 19:30:31.899495    7498 helpers.go:852] eviction manager: thresholds after resource reclaim - ignoring grace period: threshold [signal=imagefs.available, quantity=10555310237] observed 2527476Ki
Feb 14 19:30:31 dev-dal10-cr41b86b475ced4687a7ef34b5dc0513e4-w1 kubelet.service[7498]: I0214 19:30:31.899502    7498 eviction_manager.go:340] eviction manager: must evict pod(s) to reclaim ephemeral-storage
Feb 14 19:30:31 dev-dal10-cr41b86b475ced4687a7ef34b5dc0513e4-w1 kubelet.service[7498]: I0214 19:30:31.899563    7498 eviction_manager.go:358] eviction manager: pods ranked for eviction: ibm-master-proxy-static-10.94.47.199_kube-system(80da011274014ad7857773c0958a6493), calico-node-tffxn_kube-system(ed60609e-308e-11e9-84a6-021852c16f2c), ibm-kube-fluentd-9jn2q_kube-system(6d4b96f0-303b-11e9-84a6-021852c16f2c), ibm-keepalived-watcher-hcwb4_kube-system(db601843-308e-11e9-84a6-021852c16f2c)
Feb 14 19:30:31 dev-dal10-cr41b86b475ced4687a7ef34b5dc0513e4-w1 kubelet.service[7498]: E0214 19:30:31.899597    7498 eviction_manager.go:548] eviction manager: cannot evict a critical static pod ibm-master-proxy-static-10.94.47.199_kube-system(80da011274014ad7857773c0958a6493)
Feb 14 19:30:32 dev-dal10-cr41b86b475ced4687a7ef34b5dc0513e4-w1 kubelet.service[7498]: I0214 19:30:32.188866    7498 eviction_manager.go:563] eviction manager: pod calico-node-tffxn_kube-system(ed60609e-308e-11e9-84a6-021852c16f2c) is evicted successfully
Feb 14 19:30:32 dev-dal10-cr41b86b475ced4687a7ef34b5dc0513e4-w1 kubelet.service[7498]: I0214 19:30:32.188876    7498 eviction_manager.go:187] eviction manager: pods calico-node-tffxn_kube-system(ed60609e-308e-11e9-84a6-021852c16f2c) evicted, waiting for pod to be cleaned up
```
Enabling ExperimentalCriticalPodAnnotation=true makes it not fail and detect properly

**Comment by rtheis** [comment]
Setting `spec.priority` on the static pod also works.  It appears that a static pod doesn't get `spec.priority` set based on `spec.priorityClassName`.  If this is expected then it should be documented.  Although it seems like the priority class name should be honored for static pods.

**Comment by rtheis** [comment]
I've confirmed that the problem occurs when a static pod has `spec.priorityClassName` set to `system-node-critical` but does not set `spec.priority` accordingly.  `kubelet` doesn't see the priority as being set and thus thinks the pod isn't critical which ultimately causes the eviction.  I would have expected `kubelet` to translate `spec.priorityClassName` to `spec.priority`.

**Comment by pravinsinghal** [comment]
@rtheis  how does one  set it on API server started as a static pod ?

**Comment by Huang-Wei** [comment]
The code fix is available in master branch right now, which means it will be auto included in the upcoming 1.14 release.

For lower releases, you need to wait for a while before cherrypick PRs get approved and merged.

**Comment by Huang-Wei** [comment]
Cherrypick PRs have been created and waiting for approval.

- 1.13 - #74994
- 1.12 - #74995
- 1.11 - #74996

**Comment by namreg** [comment]
@Huang-Wei Your patch does not work for me. The case in that you completely replaces static pods with the mirror pods. But static pods contains important annotations such as `kubernetes.io/config.source`. For static pods is 'file', for mirror - `api` As a result, eviction manager supposes that pod is not static.

**Comment by Huang-Wei** [comment]
> For static pods is `file`, for mirror - `api`

I doubt it. A mirror pod is also annotated with `file`:

```yaml
apiVersion: v1
kind: Pod
metadata:
  annotations:
    kubernetes.io/config.hash: fb1b7ae649c5fc1b45ebbe29e456f739
    kubernetes.io/config.mirror: fb1b7ae649c5fc1b45ebbe29e456f739
    kubernetes.io/config.seen: "2019-03-07T02:51:25.531539038-06:00"
    kubernetes.io/config.source: file
  creationTimestamp: "2019-03-07T08:51:26Z"
```

**Comment by namreg** [comment]
@Huang-Wei When I get static pod through `kubectl` I see `kubernetes.io/config.source: file`. But when I try to debug, I see that pod has annotation `kubernetes.io/config.source: api`

**Comment by Huang-Wei** [comment]
@namreg Thanks for the info. Then we are sort of in a dilemma - on one hand annotation is needed to check if it's a static pod, on the other hand we also need the corresponding mirror pod to get the priority value.

I will provide a supplemental fix later.

**Comment by rtheis** [comment]
@Huang-Wei can we please hold on those cherry-picks until the supplemental fix is also available?

**Comment by Huang-Wei** [comment]
@rtheis sure.

**Comment by Huang-Wei** [comment]
Here is a detailed explanation on why previous PR passed the e2e test:

- As @namreg [pointed out](https://github.com/kubernetes/kubernetes/issues/73572#issuecomment-470426246), the critical static pod failed on IsStaticPod() check (because of mirror pod is tagged as "kubernetes.io/config.source: api")
- And it's reasonable that it's being marked as "to be deleted":
    ```
    I0307 18:09:26.138497   31769 eviction_manager.go:564] eviction manager: pod static-web-127.0.0.1_kube-system(513a7ec5-4136-11e9-b8f7-0cc47a8f8932) is evicted successfully
    ```
- However, as @ravisantoshgudimetla [observed](https://github.com/kubernetes/kubernetes/pull/74222#issuecomment-468271473), there is some weird log showing:
    ```
    I0307 18:09:27.138643   31769 kubelet_pods.go:918] Pod "static-web-127.0.0.1_kube-system(513a7ec5-4136-11e9-b8f7-0cc47a8f8932)" is terminated, but some containers are still running
    ```
- It's because the pod we passed onto pkg/kubelet/pod_workers.go#killPodNow() is a mirror pod, and it's deleted (or probably a dummy delete) by container runtime **without any error**. To take an analogy, it's sort of like a "controllee" (mirror pod) gets deleted, but "controller" (static pod) is still alive, so it causes data inconsistency, and the above error msg popped up.
- And in user's perspective, the static pod is "not" evicted, then it "looks" like previous PR works.

**Comment by Huang-Wei** [comment]
Status of cherry pick PRs:

- 1.13 - #74994 - merged
- 1.12 - #74995 - merged
- 1.11 - #74996

**Comment by Huang-Wei** [comment]
/close

As all PRs have been cherrypicked.

**Comment by rushtehrani** [comment]
When will this be released for 1.11.x?  We are getting this same issue with `kube-controller-manager` and believe it's related:  
```
The node was low on resource: ephemeral-storage. Container kube-controller-manager was using 427052188Ki, which exceeds its request of 0.
```

The api log:
```
Apr 23 11:10:16 k8s-cluster-0-master-1-nnz2 kubelet[6925]: I0423 11:10:16.133148    6925 eviction_manager.go:358] e
viction manager: pods ranked for eviction: kube-controller-manager-k8s-cluster-0-master-1-nnz2_kube-system(194e9117
f410bf5efeaf9147bb8093b2), fluentd-xzs8p_kube-logging(cf772757-655f-11e9-aa80-42010a8a0004), cluster-autoscaler-66d
4b49994-tmkgf_kube-system(d36533c6-65a2-11e9-aa80-42010a8a0004), kube-apiserver-k8s-cluster-0-master-1-nnz2_kube-sy
stem(870e41be69a61bb0473142f23af04a0d), kubernetes-dashboard-5dd89b9875-qffnm_kube-system(cf8f43e3-655f-11e9-aa80-4
2010a8a0004), etcd-k8s-cluster-0-master-1-nnz2_kube-system(0255c1dc532d6fc35547b148e51912ad), coredns-78fcdf6894-bc
84h_kube-system(cf901c4a-655f-11e9-aa80-42010a8a0004), coredns-78fcdf6894-8hdvm_kube-system(cf8f53ef-655f-11e9-aa80
-42010a8a0004), kube-scheduler-k8s-cluster-0-master-1-nnz2_kube-system(6b23efd48c0db7adcad38f9d0ab8343f), calico-no
de-bh72k_kube-system(cf7c111a-655f-11e9-aa80-42010a8a0004), kube-proxy-tm9gd_kube-system(cf794cf0-655f-11e9-aa80-42
010a8a0004)
Apr 23 11:10:18 k8s-cluster-0-master-1-nnz2 kubelet[6925]: I0423 11:10:18.012624    6925 eviction_manager.go:563] e
viction manager: pod kube-controller-manager-k8s-cluster-0-master-1-nnz2_kube-system(194e9117f410bf5efeaf9147bb8093
b2) is evicted successfully
Apr 23 11:10:18 k8s-cluster-0-master-1-nnz2 kubelet[6925]: I0423 11:10:18.012655    6925 eviction_manager.go:187] e
viction manager: pods kube-controller-manager-k8s-cluster-0-master-1-nnz2_kube-system(194e9117f410bf5efeaf9147bb809
3b2) evicted, waiting for pod to be cleaned up
Apr 23 11:10:39 k8s-cluster-0-master-1-nnz2 kubelet[6925]: I0423 11:10:39.014025    6925 eviction_manager.go:400] e
viction manager: pods kube-controller-manager-k8s-cluster-0-master-1-nnz2_kube-system(194e9117f410bf5efeaf9147bb809
3b2) successfully cleaned up
```

**Comment by Huang-Wei** [comment]
Maybe in v1.11.10 (but I didn't see it in v1.11.10-beta.0 yet)

**Comment by rushtehrani** [comment]
Ok, I assume this is pretty high priority if it's causing `kube-controller-manager` to be evicted?

**Comment by Huang-Wei** [comment]
Yes.

Prior to get it in 1.11.x, you can work it around by manually setting the priority value (not only setting priority name) to the control plane pods (apiserver, controller-manager and scheduler).

**Comment by rushtehrani** [comment]
I think the priority value is correct:

```
  priority: 2000000000
  priorityClassName: system-cluster-critical
``

**Comment by Huang-Wei** [comment]
Did you get that from `kubectl`? If so, that's not sufficient. You should work around by hardcoding it into your static pod yaml file.

**Comment by rushtehrani** [comment]
I did get it from `kubectl`, will try hardcoding. Thanks!

**Comment by aaronbbrown** [comment]
> I did get it from `kubectl`, will try hardcoding. Thanks!

@rushtehrani Were you successful ☝️?  I'm running into this same problem (unfortunately on a 1.10 cluster, which doesn't have this patch) and received the error:

```
Jun 17 07:48:21 kube-node-85cb1f5 kubelet[28566]: E0617 07:48:21.675632   28566 kubelet.go:1626] Failed creating a mirror pod for "kube-proxy-kube-node-85cb1f5_kube-system(df63a286b3fdffec236a3bad70201c1e)": pods "kube-proxy-kube-node-85cb1f5" is forbidden: the integer value of priority must not be provided in pod spec. Priority admission controller populates the value from the given PriorityClass name
```

**Comment by droidls** [comment]
> When will this be released for 1.11.x? We are getting this same issue with `kube-controller-manager` and believe it's related:
> 
> ```
> The node was low on resource: ephemeral-storage. Container kube-controller-manager was using 427052188Ki, which exceeds its request of 0.
> ```
> 
> The api log:
> 
> ```
> Apr 23 11:10:16 k8s-cluster-0-master-1-nnz2 kubelet[6925]: I0423 11:10:16.133148    6925 eviction_manager.go:358] e
> viction manager: pods ranked for eviction: kube-controller-manager-k8s-cluster-0-master-1-nnz2_kube-system(194e9117
> f410bf5efeaf9147bb8093b2), fluentd-xzs8p_kube-logging(cf772757-655f-11e9-aa80-42010a8a0004), cluster-autoscaler-66d
> 4b49994-tmkgf_kube-system(d36533c6-65a2-11e9-aa80-42010a8a0004), kube-apiserver-k8s-cluster-0-master-1-nnz2_kube-sy
> stem(870e41be69a61bb0473142f23af04a0d), kubernetes-dashboard-5dd89b9875-qffnm_kube-system(cf8f43e3-655f-11e9-aa80-4
> 2010a8a0004), etcd-k8s-cluster-0-master-1-nnz2_kube-system(0255c1dc532d6fc35547b148e51912ad), coredns-78fcdf6894-bc
> 84h_kube-system(cf901c4a-655f-11e9-aa80-42010a8a0004), coredns-78fcdf6894-8hdvm_kube-system(cf8f53ef-655f-11e9-aa80
> -42010a8a0004), kube-scheduler-k8s-cluster-0-master-1-nnz2_kube-system(6b23efd48c0db7adcad38f9d0ab8343f), calico-no
> de-bh72k_kube-system(cf7c111a-655f-11e9-aa80-42010a8a0004), kube-proxy-tm9gd_kube-system(cf794cf0-655f-11e9-aa80-42
> 010a8a0004)
> Apr 23 11:10:18 k8s-cluster-0-master-1-nnz2 kubelet[6925]: I0423 11:10:18.012624    6925 eviction_manager.go:563] e
> viction manager: pod kube-controller-manager-k8s-cluster-0-master-1-nnz2_kube-system(194e9117f410bf5efeaf9147bb8093
> b2) is evicted successfully
> Apr 23 11:10:18 k8s-cluster-0-master-1-nnz2 kubelet[6925]: I0423 11:10:18.012655    6925 eviction_manager.go:187] e
> viction manager: pods kube-controller-manager-k8s-cluster-0-master-1-nnz2_kube-system(194e9117f410bf5efeaf9147bb809
> 3b2) evicted, waiting for pod to be cleaned up
> Apr 23 11:10:39 k8s-cluster-0-master-1-nnz2 kubelet[6925]: I0423 11:10:39.014025    6925 eviction_manager.go:400] e
> viction manager: pods kube-controller-manager-k8s-cluster-0-master-1-nnz2_kube-system(194e9117f410bf5efeaf9147bb809
> 3b2) successfully cleaned up
> ```

I see this as well but the actual disk is not even full.
